libraries/AP_Motors/AP_MotorsHeli_Swash.cpp

```diff
@@ -0,0 +1,200 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdlib.h>
+#include <AP_HAL/AP_HAL.h>
+
+#include "AP_MotorsHeli_Swash.h"
+
+extern const AP_HAL::HAL& hal;
+
+const AP_Param::GroupInfo SwashInt16Param::var_info[] = {
+
+    // @Param: ENABLE
+    // @DisplayName: Enable settings for H3
+    // @Description: Automatically set when H3 swash type is selected. Should not be set manually.
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLE", 1, SwashInt16Param, enable, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: SV1_POS
+    // @DisplayName: servo1_pos
+    // @Description: servo 1 position
+    // @Range: -180 180
+    // @Units: deg
+    // @User: Advanced
+    AP_GROUPINFO("SV1_POS", 2, SwashInt16Param, servo1_pos, -60),
+
+    // @Param: SV2_POS
+    // @DisplayName: servo2_pos
+    // @Description: servo 2 position
+    // @Range: -180 180
+    // @Units: deg
+    // @User: Advanced
+    AP_GROUPINFO("SV2_POS", 3, SwashInt16Param, servo2_pos, 60),
+
+    // @Param: SV3_POS
+    // @DisplayName: servo3_pos
+    // @Description: servo 3 position
+    // @Range: -180 180
+    // @Units: deg
+    // @User: Advanced
+    AP_GROUPINFO("SV3_POS", 4, SwashInt16Param, servo3_pos, 180),
+    
+    // @Param: PHANG
+    // @DisplayName: Swashplate Phase Angle Compensation
+    // @Description: Only for H3 swashplate.  If pitching the swash forward induces a roll, this can be correct the problem
+    // @Range: -30 30
+    // @Units: deg
+    // @User: Advanced
+    // @Increment: 1
+    AP_GROUPINFO("PHANG", 5, SwashInt16Param, phase_angle, 0),
+   
+    AP_GROUPEND
+};
+
+/*
+  a manual swashplate definition with enable and servo position parameters - constructor
+ */
+SwashInt16Param::SwashInt16Param(void)
+{
+    AP_Param::setup_object_defaults(this, var_info);    
+}
+
+// CCPM Mixers - calculate mixing scale factors by swashplate type
+void AP_MotorsHeli_Swash::calculate_roll_pitch_collective_factors()
+{
+    if (_swash_type == SWASHPLATE_TYPE_H1) {
+        // CCPM mixing not used
+        _collectiveFactor[CH_1] = 0;
+        _collectiveFactor[CH_2] = 0;
+        _collectiveFactor[CH_3] = 1;
```

Comment:
No need to worry about the collectiveFactor[CH_4] not being set even in the case that a users switches the swash type between an H1 type and an H4?  I suspect it's no problem because once switched it likely doesn't use the collectiveFactor no matter what it is but though i'd ask..

Reply:
I actually wrote the mixers and Bill just pulled it from my code. No, it doesn't make a difference because the H1 is a special case for three-servo mechanically mixed helicopters. So once a user switches to a CCPM swash type it is not used. Interestingly, the terminology from the mechanically mixed helicopters (pitch, aileron, elevator servos) carried over into CCPM setups when electronic mixing first appeared. ArduPilot, for whatever reason, departed from this which makes the code confusing compared to all other FBL setups. Even in full-size the links to the swashplate are called "pitch, aileron and elevator links". What is called "collective" in ArduPilot is actually the pitch servo or link, and referred to as "pulling pitch" in helicopters. What ArduPilot calls "pitch" is actually the elevator servo or link. So in writing the mixers I have to deal with the ArduPilot concept and try to apply it to the world of helicopters as helicopter engineering is understood so it works with the rest of the ArduPilot code.

Hope that makes sense. When I did this I looked at how I could change it to conventional helicopter setup. But it would require a complete re-write of the Motors class. So I just made it work with ArduPilot's strange concept and separated the collective (which is really pitch) from the rest of the mixers so it didn't have to be re-stated numerous times as was done in the original code :-)